src/semantics.rs

```diff
@@ -13,55 +14,6 @@ pub trait FieldCounter {
     fn remove_from_counter(&self, map: &mut HashMap<String, usize>);
 }
 
-impl FieldCounter for Expression {
```

Comment:
Why not keep `FieldCounter` operating on the AST, not `Lir` for a few reasons:

1. It allows us flexibility for changing the VM implementation
2. It clearly separating the responsibility of VM and Parser, which results structurally better code
3. Semantics does not depends on `Lir` implementation details

Could you explain if there are anything that I'm missing that might result in different decision being made?

Reply:
In my opinion, the `FieldCounter` is not a feature of a DSL spec, it is a feature of its usage (atc-router), so we should not implement it on AST as the responsibility of AST is to reflect the spec of the DSL

The `atc-router` must have the ability to do incremental rebuild, so we need the `FieldCounter`, this is the requirement of the `atc-router`, which is not in the spec of `atc-lang` yet.

The `Lir` is a part of the implementation for `atc-lang`, which is used by the `atc-router`. For the best practice of type-level programming, the `FieldCounter` should be implemented by `atc-router`, or `atc-vm` that is used by the `atc-router`.

Currently, we haven't split the `atc-router` into multiple crates, I think we should do it the make the responsibility and dataflow more clear.

Reply:
Deriving further.

For the use case that needs incremental rebuild, we should choose a VM that supports incremental rebuild and sacrifices the opportunities of optimization.

For the use case that doesn't need it, we should also choose a VM that can do more optimization by sacrificing incremental build

Reply:
Instead of type-level programming, from the implementation perspective, we have lost the old ASTs when adding/removing a matcher, so keeping the `FieldCounter` for AST doesn't help to rebuild, so we removed the dead code yo reduce the maintenance burden.

Reply:
@ADD-SP For our use case, sacrefising incremental rebuild is basically a no-go. But having field information at parse time is not only critical for Kong, but also for a lot of projects that use this crate. I recommend us to consider more carefully before reaching a conclusion.